File: combinations/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def helper(start, path):
            if len(path) + 1 == k: #Add the final element of possible paths
                for i in range(start+1, n+1):
                    answer.append(path + [i])
            else:
                for i in range(start+1, n+1): #Build the path
                    helper(i, path + [i])

        answer = []
        helper(0, [])
        return answer

File: combinations/combinations.java
class Solution 
{
    public List<List<Integer>> combine(int n, int k) 
    {
        List<List<Integer>> answer = new ArrayList<>();
        if (n < 1) // Nothing to do
            return answer;
        
        combinations(answer, new ArrayList<Integer>(), 1, n, k);
        return answer;
    }
    
    public void combinations(List<List<Integer>> answer, List<Integer> current, int start, int end, int k)
    {
        if (k == 0) // Combination complete
        {
            answer.add(new ArrayList<Integer>(current));
            return;
        }
        
        for (int i = start; i <= end; i++) // Generate combinations
        {
            current.add(i);
            combinations(answer, current, i+1, end, k-1);
            current.remove(current.size()-1);
        }
    }
}